### DISCUSSION TITLE
Usage of external storage adapter

### DISCUSSION THREAD

**Discussion by marcoschierhorn** [shyim/shopware#37] [Q&A]
First of all, thanks for your great work!
Could you elaborate on how to use an external storage adapter in the configuration (docker-compose.yml) file?

/public/*  Use external storage adapter, see Shopware documentation

**Comment by shyim** [comment]
https://developer.shopware.com/docs/guides/hosting/infrastructure/filesystem

You have to mount the configuration to /var/www/html/custom/config/{choose name}.yml

**Reply by marcoschierhorn** [reply]
I have tried to just set it in "/var/www/html/config/packages/shopware.yaml":

`shopware:
  auto_update:
    enabled: false
  filesystem:
      public:
        type: "google-storage"
        config:
           bucket: "shopware-gke-test"
           projectId: "test-shopware-on-gcp"
           keyFilePath: "/var/www/html/config/packages/shopware-gcp.json"`

Now all images are uploaded to GCS, but unfortunately they are not displayed at the shop frontend.
Do I also need to change something in docker-compose.yml ?

**Reply by shyim** [reply]
Should be enough. public filesystem only considers uploaded images, not assets like css.

Maybe cache

**Reply by marcoschierhorn** [reply]
Great, thank you! Can this also be directly defined in the docker-compose.yml or do I need to apply the changes to shopware.yaml always? 
The files are also uploaded correctly to Google Cloud Filestorage (I've installed the Demo content), but at the frontend I only get 404 errors for the images. Do I need to change something else (some kind of routing) or apply specific definitions at docker-compose.yml?

**Reply by shyim** [reply]
Currently not as the shopware.yaml is huge what it has anything. 

If you change the storage, you have to move by own the files. Is that what you mean?

**Comment by marcoschierhorn** [comment]
Ah ok, the I'll just change it in shopware.yaml.

I started with an empty shop. The applied the changes to shopware.yml and
installed and activated the Demo Shop plugin. The plugin routine then
installed the products in the product catalog and uploaded all
media/product images to Google Cloud Storage. The files in GCS are fine,
but in the backend and frontend no images are shown. I only get 404 errors
for the product images.

Shyim ***@***.***> schrieb am So., 22. Mai 2022, 21:08:

**Reply by marcoschierhorn** [reply]
Good morning, I have also tried to use "url" before the adpater configuration:

`shopware:
  auto_update:
    enabled: false
  filesystem:
      public:
        url: "https://storage.googleapis.com/shopware-gke-test"
        type: "google-storage"
        config:
           bucket: "shopware-gke-test"
           projectId: "test-shopware-on-gcp"
           keyFilePath: "/var/www/html/config/packages/shopware-gcp.json"`

But I always get this error message than:

Fatal error: Uncaught Symfony\Component\Config\Definition\Exception\InvalidConfigurationException: Unrecognized option "url" under "shopware.filesystem.public". Available options are "config", "type". in /var/www/html/vendor/symfony/config/Definition/ArrayNode.php:327 Stack trace: #0 /var/www/html/vendor/symfony/config/Definition/BaseNode.php(382): Symfony\Component\Config\Definition\ArrayNode->normalizeValue(Array) #1 /var/www/html/vendor/symfony/config/Definition/ArrayNode.php(292): Symfony\Component\Config\Definition\BaseNode->normalize(Array) #2 /var/www/html/vendor/symfony/config/Definition/BaseNode.php(382): Symfony\Component\Config\Definition\ArrayNode->normalizeValue(Array) #3 /var/www/html/vendor/symfony/config/Definition/ArrayNode.php(292): Symfony\Component\Config\Definition\BaseNode->normalize(Array) #4 /var/www/html/vendor/symfony/config/Definition/BaseNode.php(382): Symfony\Component\Config\Definition\ArrayNode->normalizeValue(Array) #5 /var/www/html/vendor/symfony/config/Definition/Processor.php(34): Symfon in /var/www/html/vendor/symfony/config/Definition/ArrayNode.php on line 327

**Reply by shyim** [reply]
Which shopware version do you use?

**Comment by marcoschierhorn** [comment]
I finally got it working with adding the cdn attributes to shopware.yaml (see below).
I have now transferred everything to Google Cloud Artifact Registry and deployed it to a cluster. Do you know how I can make Nginx work in the GKE cluster? It seems that it's not pointing to the right directory.
I'm also trying to get the Cloud SQL connection working via a Proxy.

shopware:
  auto_update:
    enabled: false
  filesystem:
      public:
        type: "google-storage"
        config:
           root: "https://storage.googleapis.com/shopware-gke-test"
           bucket: "shopware-gke-test"
           projectId: "test-shopware-on-gcp"
           keyFilePath: "/var/www/html/config/packages/shopware-gcp.json"
  cdn:
    url: "https://storage.googleapis.com/shopware-gke-test"

**Comment by marcoschierhorn** [comment]
Thanks for your great support!
Two last questions, is it possible to copy a predefined shopware.yml from the rootfs to the shopware environment via using the init/start scripts?
Can I also pass environmental variables for setting up the docker container (like PORT or DB Socket)?
